// webdriverio/build/scripts/isElementClickable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = isElementClickable;

function isElementClickable(elem) {
  if (!elem.getBoundingClientRect || !elem.scrollIntoView || !elem.contains || !elem.getClientRects || !document.elementFromPoint) {
    return false;
  }

  function getOverlappingElement(elem, context = document) {
    const elemDimension = elem.getBoundingClientRect();
    const x = elemDimension.left + elem.clientWidth / 2;
    const y = elemDimension.top + elem.clientHeight / 2;
    return context.elementFromPoint(x, y);
  }

  function getOverlappingRects(elem, context = document) {
    const elems = [];
    const rects = elem.getClientRects();
    const rect = rects[0];
    const x = rect.left + rect.width / 2;
    const y = rect.top + rect.height / 2;
    elems.push(context.elementFromPoint(x, y));
    return elems;
  }

  function getOverlappingElements(elem, context) {
    return [getOverlappingElement(elem, context), ...getOverlappingRects(elem, context)];
  }

  function isOverlappingElementMatch(elementsFromPoint, elem) {
    if (elementsFromPoint.some(elementFromPoint => elementFromPoint === elem || elem.contains(elementFromPoint))) {
      return true;
    }

    let elemsWithShadowRoot = [...new Set(elementsFromPoint)];
    elemsWithShadowRoot = elemsWithShadowRoot.filter(x => x && x.shadowRoot && x.shadowRoot.elementFromPoint);
    let shadowElementsFromPoint = [];

    for (let shadowElement of elemsWithShadowRoot) {
      shadowElementsFromPoint.push(...getOverlappingElements(elem, shadowElement.shadowRoot));
    }

    shadowElementsFromPoint = [...new Set(shadowElementsFromPoint)];
    shadowElementsFromPoint = shadowElementsFromPoint.filter(x => !elementsFromPoint.includes(x));

    if (shadowElementsFromPoint.length === 0) {
      return false;
    }

    return isOverlappingElementMatch(shadowElementsFromPoint, elem);
  }

  function isElementInViewport(elem) {
    if (!elem.getBoundingClientRect) {
      return false;
    }

    const rect = elem.getBoundingClientRect();
    const windowHeight = window.innerHeight || document.documentElement.clientHeight;
    const windowWidth = window.innerWidth || document.documentElement.clientWidth;
    const vertInView = rect.top <= windowHeight && rect.top + rect.height > 0;
    const horInView = rect.left <= windowWidth && rect.left + rect.width > 0;
    return vertInView && horInView;
  }

  function isClickable(elem) {
    return isElementInViewport(elem) && elem.disabled !== true && isOverlappingElementMatch(getOverlappingElements(elem), elem);
  }

  let scrollIntoViewFullSupport = !window.safari || !window.StyleMedia;

  if (!isClickable(elem)) {
    elem.scrollIntoView(scrollIntoViewFullSupport ? {
      block: 'nearest',
      inline: 'nearest'
    } : false);

    if (!isClickable(elem)) {
      elem.scrollIntoView({
        block: 'center',
        inline: 'center'
      });
      return isClickable(elem);
    }
  }

  return true;
}
